### PR TITLE
Add implementation of the Data Mapper Compiler Plugin

### PR DESCRIPTION
**Pull request by miyurud**
## Purpose
Currently there is no way to extract connector information that are required by data mapper service.

## Goals
Develop a new compiler extension that does the following,
- Automatically extract the connector actions and their related Types' schemas and write those to a set of JSON files
- Allow connector developers to create example data in JSON format that can be used by data mapper service
- Validate the example JSON data provided by the connector developers

## Approach
Here we have developed a Ballerina compiler extension which uses the Ballerina AST as well as sample data provided by the connector developers.

## Automation tests
 - Integration tests
We have developed multiple integration tests to validate the use of the compiler extension

## Test environment
JDK 1.8, Ballerina 1.2.X, Ubuntu 18.04
 
## Learning
[How to Extend Ballerina](https://ballerina.io/learn/how-to-extend-ballerina/)

### CODE REVIEW

File: src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        } else if (recordKeyValue.valueExpr.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+            recordKeyValue.valueExpr.accept(this);
+        } else if (recordKeyValue.valueExpr.getKind() == NodeKind.NUMERIC_LITERAL) {
+            recordKeyValue.valueExpr.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangRecordLiteral recordLiteral) {
+        for (Iterator<RecordLiteralNode.RecordField> iterator = recordLiteral.fields.iterator(); iterator.hasNext(); ) {
```
Comment (rasika) [inline src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java:129]: Can't we replace this with a simple for-each loop. eg. `for(RecordLiteralNode.RecordField field : recordLiteral.fields){...}`

File: src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public void visit(BLangSimpleVariable varNode) {
+        structureMap.put(varNode.symbol.name.toString(), varNode.symbol.type.toString());
+    }
+
+    @Override
+    public void visit(BLangRecordLiteral.BLangRecordKeyValueField recordKeyValue) {
+        if (recordKeyValue.valueExpr.getKind() == NodeKind.LITERAL) {
```
Comment (rasika) [inline src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java:114]: IMO a switch case would be more readable `switch(recordKeyValue.valueExpr.getKind()){...}`

File: src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java (hunk 1 above)
Comment (miyurud) [reply src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java]: Yes, we can use this approach. I update the code to follow this approach.

File: src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java (hunk 2 above)
Comment (miyurud) [reply src/main/java/org/ballerinax/datamapper/DataMapperStructureVisitor.java]: Yes, the readability can be improved using the switch statement. I updated the code to follow this approach.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,68 @@
-# module-ballerinax-datamapper
-A compiler extension to extract abstract representation of Ballerina connector actions and their associated types
+# Ballerina Data Mapper Extension
+A compiler extension to extract an abstract representation of Ballerina connector actions and their associated types. When compiling a
```
Comment (praneesha) [inline README.md:2]:
```suggestion
The Ballerina Data Mapper extension is a compiler extension, which extracts an abstract representation of Ballerina connector actions and the associated types. 
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,68 @@
-# module-ballerinax-datamapper
-A compiler extension to extract abstract representation of Ballerina connector actions and their associated types
+# Ballerina Data Mapper Extension
+A compiler extension to extract an abstract representation of Ballerina connector actions and their associated types. When compiling a 
+Ballerina connector project this compiler extension gets executed. The extension generates two types of files by default. Note that
```
Comment (praneesha) [inline README.md:3]:
```suggestion
This compiler extension gets executed when compiling a Ballerina connector project. The extension generates two types of files by default. Note that 
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,68 @@
-# module-ballerinax-datamapper
-A compiler extension to extract abstract representation of Ballerina connector actions and their associated types
+# Ballerina Data Mapper Extension
+A compiler extension to extract an abstract representation of Ballerina connector actions and their associated types. When compiling a 
+Ballerina connector project this compiler extension gets executed. The extension generates two types of files by default. Note that 
+these two types of files get automatically extracted from connectors hence developer does not need to write them.
```
Comment (praneesha) [inline README.md:4]:
```suggestion
these two types of files get automatically extracted from connectors. Hence, the developer does not need to write them.
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,68 @@
...
+
+## Functions File
+These files end with a file name ``<CLIENT_NAME>_functions.json`` where the `CLIENT_NAME` corresponds to the connector client name.
```
Comment (praneesha) [inline README.md:7]:
```suggestion
These files are named as ``<CLIENT_NAME>_functions.json`` in which the `CLIENT_NAME` corresponds to the connector client name.
```

File: README.md [hunk 7]
```diff
@@ -1,2 +1,68 @@
...
+When there are ``n`` clients exist in a connector, ``n``  number of functions files get created. Function files get created only if there
```
Comment (praneesha) [inline README.md:8]:
```suggestion
When there are ``n`` number of clients existing in a connector, a ``n``  number of functions files will get created. Function files get created only if there 
```

File: README.md [hunk 8]
```diff
@@ -1,2 +1,68 @@
...
+these two types of files get automatically extracted from connectors hence developer does not need to write them.
+
+## Functions File
+These files end with a file name ``<CLIENT_NAME>_functions.json`` where the `CLIENT_NAME` corresponds to the connector client name.
+When there are ``n`` clients exist in a connector, ``n``  number of functions files get created. Function files get created only if there 
+are client objects with remote function calls. These files can be used to identify the list of connector actions available in a 
+particular connector as well as to identify what types of records are used as input parameters and return types.
+
+## Schema File
+These files end with a file name ``<RECORD_NAME>_schema.json`` where the `RECORD_NAME` corresponds to a record type defined
```
Comment (praneesha) [inline README.md:13]:
```suggestion
These files are named as ``<RECORD_NAME>_schema.json`` in which the `RECORD_NAME` corresponds to a record type defined
```

File: README.md [hunk 9]
```diff
@@ -1,2 +1,68 @@
...
+within the connector module. In order for a schema file to be generated for a record type defined within a connector module, the
+record has to be one of the following,
+ 
+ 1. An input parameter or a return type from a connector action.
+ 2. A record type referred by a record which falls into the category (1).
```
Comment (praneesha) [inline README.md:18]:
```suggestion
 2. A record type referred by a record, which falls into the first category.
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,68 @@
...
+ 
+  In addition to the above two types of JSON file categories which gets automatically extracted from a connector by default,
```
Comment (praneesha) [inline README.md:20]:
```suggestion
  In addition to the above two types of JSON file categories, which get automatically extracted from a connector by default, 
```

File: README.md [hunk 11]
```diff
@@ -1,2 +1,68 @@
...
+## Schema File
+These files end with a file name ``<RECORD_NAME>_schema.json`` where the `RECORD_NAME` corresponds to a record type defined
+within the connector module. In order for a schema file to be generated for a record type defined within a connector module, the
+record has to be one of the following,
+ 
+ 1. An input parameter or a return type from a connector action.
+ 2. A record type referred by a record which falls into the category (1).
+ 
+  In addition to the above two types of JSON file categories which gets automatically extracted from a connector by default, 
+  the connector developer can also specify sample data in files having file names `<RECORD_NAME>_data.json`.
```
Comment (praneesha) [inline README.md:21]:
```suggestion
  the connector developer can also specify sample data in the `<RECORD_NAME>_data.json` files.
```

File: README.md [hunk 12]
```diff
@@ -1,2 +1,68 @@
...
+    "<ORG_NAME>/<MODULE_NAME>:<VERSION_NUMBER>:<RECORD_NAME>": [
+        <SAMPLE_JSON_INSTANCE_1>,
+        <SAMPLE_JSON_INSTANCE_2>,
+        <SAMPLE_JSON_INSTANCE_3>,
+        ...,
+        <SAMPLE_JSON_INSTANCE_N>
+    ]
+}
+```
+The `<SAMPLE_JSON_INSTANCE_i>` should have a structure as follows where the JSON
```
Comment (praneesha) [inline README.md:37]:
```suggestion
The `<SAMPLE_JSON_INSTANCE_i>` should have a structure as follows in which the JSON 
```

File: README.md [hunk 13]
```diff
@@ -1,2 +1,68 @@
...
+        <SAMPLE_JSON_INSTANCE_1>,
+        <SAMPLE_JSON_INSTANCE_2>,
+        <SAMPLE_JSON_INSTANCE_3>,
+        ...,
+        <SAMPLE_JSON_INSTANCE_N>
+    ]
+}
+```
+The `<SAMPLE_JSON_INSTANCE_i>` should have a structure as follows where the JSON 
+instance has `M` number of attributes,
```
Comment (praneesha) [inline README.md:38]:
```suggestion
instance has `M` number of attributes.
```

File: README.md [hunk 14]
```diff
@@ -1,2 +1,68 @@
...
+    "<ATTRIBUTE_NAME_3>" : "<ATTRIBUTE_VALUE_3>",
+    ...,
+    "<ATTRIBUTE_NAME_M>" : "<ATTRIBUTE_VALUE_M>"
+}
+```
+
+An `<ATTRIBUTE_VALUE_j>` can be a simple literal (string/numeric value), 
+`<SAMPLE_JSON_INSTANCE_j>`, an array, or `null`.
+
+Unlike functions files and schema files which automatically gets extracted,
```
Comment (praneesha) [inline README.md:53]:
```suggestion
Unlike functions files and schema files, which automatically get extracted
```

File: README.md [hunk 15]
```diff
@@ -1,2 +1,68 @@
...
+    ...,
+    "<ATTRIBUTE_NAME_M>" : "<ATTRIBUTE_VALUE_M>"
+}
+```
+
+An `<ATTRIBUTE_VALUE_j>` can be a simple literal (string/numeric value), 
+`<SAMPLE_JSON_INSTANCE_j>`, an array, or `null`.
+
+Unlike functions files and schema files which automatically gets extracted, 
+from the Ballerina Abstract Syntax Tree (AST), data files are user specified JSON files.
```
Comment (praneesha) [inline README.md:54]:
```suggestion
from the Ballerina Abstract Syntax Tree (AST), data files are user-specified JSON files. 
```

File: README.md [hunk 16]
```diff
@@ -1,2 +1,68 @@
...
+    "<ATTRIBUTE_NAME_M>" : "<ATTRIBUTE_VALUE_M>"
+}
+```
+
+An `<ATTRIBUTE_VALUE_j>` can be a simple literal (string/numeric value), 
+`<SAMPLE_JSON_INSTANCE_j>`, an array, or `null`.
+
+Unlike functions files and schema files which automatically gets extracted, 
+from the Ballerina Abstract Syntax Tree (AST), data files are user specified JSON files. 
+Hence the Data Mapper compiler extension conducts a validation of the content of the data JSON files.
```
Comment (praneesha) [inline README.md:55]:
```suggestion
Hence, the Data Mapper compiler extension conducts a validation of the content of the data JSON files.
```

File: README.md [hunk 17]
```diff
@@ -1,2 +1,68 @@
...
+The validation process checks whether the data json file's attribute names are equivalent to the
```
Comment (praneesha) [inline README.md:56]:
```suggestion
The validation process checks whether the data JSON file's attribute names are equivalent to the 
```

File: README.md [hunk 18]
```diff
@@ -1,2 +1,68 @@
...
+
+An `<ATTRIBUTE_VALUE_j>` can be a simple literal (string/numeric value), 
+`<SAMPLE_JSON_INSTANCE_j>`, an array, or `null`.
+
+Unlike functions files and schema files which automatically gets extracted, 
+from the Ballerina Abstract Syntax Tree (AST), data files are user specified JSON files. 
+Hence the Data Mapper compiler extension conducts a validation of the content of the data JSON files.
+The validation process checks whether the data json file's attribute names are equivalent to the 
+attribute names used in their corresponding types. Furthermore, it checks whether the same number 
+of attributes exists between the record type and its corresponding data JSON. In addition to these
```
Comment (praneesha) [inline README.md:58]:
```suggestion
of attributes exist between the record type and its corresponding data JSON file. In addition to these 
```

File: README.md [hunk 19]
```diff
@@ -1,2 +1,68 @@
...
+`<SAMPLE_JSON_INSTANCE_j>`, an array, or `null`.
+
+Unlike functions files and schema files which automatically gets extracted, 
+from the Ballerina Abstract Syntax Tree (AST), data files are user specified JSON files. 
+Hence the Data Mapper compiler extension conducts a validation of the content of the data JSON files.
+The validation process checks whether the data json file's attribute names are equivalent to the 
+attribute names used in their corresponding types. Furthermore, it checks whether the same number 
+of attributes exists between the record type and its corresponding data JSON. In addition to these 
+validations, the compiler extension also extracts sample data nested within a particular data JSON 
+to separate top level data JSON files.
```
Comment (praneesha) [inline README.md:60]:
```suggestion
to separate top-level data JSON files.
```

File: README.md [hunk 20]
```diff
@@ -1,2 +1,68 @@
...
+
+The extracted json files get written to the resources directory within the connector module's
```
Comment (praneesha) [inline README.md:62]:
```suggestion
The extracted JSON files get written to the resources directory within the connector module's 
```

File: README.md [hunk 21]
```diff
@@ -1,2 +1,68 @@
...
+Hence the Data Mapper compiler extension conducts a validation of the content of the data JSON files.
+The validation process checks whether the data json file's attribute names are equivalent to the 
+attribute names used in their corresponding types. Furthermore, it checks whether the same number 
+of attributes exists between the record type and its corresponding data JSON. In addition to these 
+validations, the compiler extension also extracts sample data nested within a particular data JSON 
+to separate top level data JSON files.
+
+The extracted json files get written to the resources directory within the connector module's 
+`src` folder. However, the functions and schema JSON files are not required to be committed to a
+Github repository since they get automatically generated from the connector's Ballerina code. However,
```
Comment (praneesha) [inline README.md:64]:
```suggestion
Github repository since they get generated automatically from the connector's Ballerina code. However, 
```

**Pull request merged**